lib/ansible/modules/cloud/xenserver/xenserver_guest.py

```diff
@@ -0,0 +1,1510 @@
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state in ["paused", "running"] and self.module.check_mode:
+                return True
+
+            # hard_restart will restart VM regardless of current state.
+            if vm_power_state in ["paused", "running"]:
+                self.xapi_session.xenapi.VM.hard_reboot(self.vm_ref)
+                self.vm_params['power_state'] = "running"
+                changed_state = True
+            else:
+                self.module.fail_json(msg="Cannot restart VM in state '%s'!" % vm_power_state)
+        except XenAPI.Failure as f:
+            self.module.fail_json(msg="XAPI ERROR: %s" % f.details)
+
+        return changed_state
+
+    def shutdownguest(self):
+        # Safety check.
+        if self.vm_ref is None:
+            self.module.fail_json(msg="Called shutdownguest on non existing VM!")
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest shutdown.
+            if vm_power_state == "running":
+                if self.module.params['state_change_timeout'] == 0:
+                    self.xapi_session.xenapi.VM.clean_shutdown(self.vm_ref)
+                else:
+                    task_ref = self.xapi_session.xenapi.Async.VM.clean_shutdown(self.vm_ref)
+                    task_result = wait_for_task(self.module, task_ref, self.module.params['state_change_timeout'])
+
+                    if task_result:
+                        self.module.fail_json(msg="Guest shutdown task failed - '%s'!" % task_result)
+
+                self.vm_params['power_state'] = "halted"
+                changed_state = True
+            else:
+                self.module.fail_json(msg="VM must be in running state for guest shutdown! Current VM state is '%s'." % vm_power_state)
+        except XenAPI.Failure as f:
+            self.module.fail_json(msg="XAPI ERROR: %s" % f.details)
+
+        return changed_state
+
+    def rebootguest(self):
+        # Safety check.
+        if self.vm_ref is None:
+            self.module.fail_json(msg="Called rebootguest on non existing VM!")
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest reboot
```

Comment:
I would recommend to have a separate module for managing power state of VMs. We have seen issues in vmware_guest module related to this and created new module called vmware_guest_powerstate ?

Reply:
Hmmm. If that is a general course of action with vmware_guest module, I can go with that too. So is power state management for vmware_guest module planned for removal? Also, if I split power state management to separate module, should I make that module part of this pull request?

Reply:
There is no plan to remove power state management from vmware_guest for backward compatibility reason. But, for this module I would say it will be good to manage powerstate separately. Submit a separate power state module as per - https://docs.ansible.com/ansible/2.5/dev_guide/developing_modules_in_groups.html#subsequent-prs

Reply:
Hmmm, with current design a user can specify power state "poweredon" for non existing VM. In that case, the module will deploy a new VM based on module parameters and run it automatically after deploy. I wouldn't like for users to be forced to use two separate modules and two tasks in a playbook to accomplish the same, especially if they can do the same with vmware_guest module and will be able to do it for foreseeable future. I think that at least "poweredon" and maybe "poweredoff" states should be supported by this module - it removes the clutter from playbook. Other states are much less relevant.

Reply:
ok.

Reply:
Great. Then my plan is to split power state management functionality from xenserver_guest module and put it in xenserver module util so that it can be shared by multiple modules.